File: benchmarks/blue_noise.py
# https://blog.demofox.org/2017/10/20/generating-blue-noise-sample-points-with-mitchells-best-candidate-algorithm/

import sys
import time
import json
import random
import kdtree

def run_bench(n):
    bounds = kdtree.Pointd(1000, 1000)
    metric = kdtree.ToroidalL2(kdtree.L2(), bounds)
    tree = kdtree.KDTreed()

    def rand_p():
        return (random.uniform(0, bounds.x), random.uniform(0, bounds.y))

    start = time.perf_counter()
    tree.insert(rand_p(), 0)

    total_query_ms = 0
    query_count = 0

    for i in range(1, n):
        num_candidates = i + 1
        best_p = None
        max_sq_dist = -1

        for _ in range(num_candidates):
            p = rand_p()
            q_start = time.perf_counter()
            closest = tree.find_closest(p, metric=metric)
            total_query_ms += (time.perf_counter() - q_start) * 1000
            query_count += 1

            dx = abs(p[0] - closest.p.x)
            dy = abs(p[1] - closest.p.y)
            dx = min(dx, bounds.x - dx)
            dy = min(dy, bounds.y - dy)
            sq_dist = dx*dx + dy*dy

            if sq_dist > max_sq_dist:
                max_sq_dist = sq_dist
                best_p = p
        tree.insert(best_p, i)

    total_ms = (time.perf_counter() - start) * 1000

    return [
        {"test": "Blue Noise Total", "implementation": "Python", "n": n, "time_ms": total_ms, "iters": 1},
        {"test": "Blue Noise Search", "implementation": "Python", "n": n, "time_ms": total_query_ms, "iters": query_count}
    ]

if __name__ == "__main__":
    sizes = [500, 1000, 2000]
    all_results = []
    for n in sizes:
        all_results.extend(run_bench(n))
    print(json.dumps(all_results, indent=2))

File: benchmarks/blue_noise.cpp
// https://blog.demofox.org/2017/10/20/generating-blue-noise-sample-points-with-mitchells-best-candidate-algorithm/

#include <iostream>
#include <vector>
#include <chrono>
#include <random>
#include <algorithm>
#include "kdtree.h"

using namespace kdtree;

void run_bench(int n, bool first) {
    Pointd bounds(1000.0, 1000.0);
    Toroidal<L2sq, double> metric{bounds};
    KDTreed tree;

    std::mt19937_64 gen(42);
    std::uniform_real_distribution<double> dist_x(0.0, bounds.x);
    std::uniform_real_distribution<double> dist_y(0.0, bounds.y);
    auto rand_p = [&]() { return Pointd(dist_x(gen), dist_y(gen)); };

    auto start = std::chrono::high_resolution_clock::now();
    tree.insert(rand_p(), 0);

    double total_query_ms = 0;
    int query_count = 0;

    for (int i = 1; i < n; ++i) {
        int num_candidates = i + 1;
        Pointd best_p;
        double max_sq_dist = -1.0;

        for (int j = 0; j < num_candidates; ++j) {
            Pointd p = rand_p();
            auto q_start = std::chrono::high_resolution_clock::now();
            auto closest = tree.find_closest(p, metric, -1.0);
            auto q_end = std::chrono::high_resolution_clock::now();

            total_query_ms += std::chrono::duration<double, std::milli>(q_end - q_start).count();
            query_count++;

            double d2 = metric.dist(p, closest->p);
            if (d2 > max_sq_dist) {
                max_sq_dist = d2;
                best_p = p;
            }
        }
        tree.insert(best_p, i);
    }

    auto end = std::chrono::high_resolution_clock::now();
    double total_ms = std::chrono::duration<double, std::milli>(end - start).count();

    if (!first) std::cout << ",\n";
    std::cout << "  {\"test\": \"Blue Noise Total\", \"implementation\": \"C++\", \"n\": " << n << ", \"time_ms\": " << total_ms << ", \"iters\": 1},\n";
    std::cout << "  {\"test\": \"Blue Noise Search\", \"implementation\": \"C++\", \"n\": " << n << ", \"time_ms\": " << total_query_ms << ", \"iters\": " << query_count << "}";
}

int main() {
    std::vector<int> sizes = {500, 1000, 2000};
    std::cout << "[\n";
    for (size_t i = 0; i < sizes.size(); ++i) {
        run_bench(sizes[i], i == 0);
    }
    std::cout << "\n]\n";
    return 0;
}
